File: examples/crypto/keccak256/commitment-schemes.ts
import { Keccak256 } from "../../../src/crypto/Keccak256/index.js";
import { Hex } from "../../../src/primitives/Hex/index.js";

/**
 * Cryptographic Commitment Schemes
 *
 * Demonstrates using Keccak256 for commitments:
 * - Commit-reveal schemes
 * - Secret commitments with nonces
 * - Timestamped commitments
 * - Merkle proofs
 * - Verifiable random functions
 */

console.log("=== Cryptographic Commitment Schemes ===\n");

// 1. Basic Commit-Reveal
console.log("1. Basic Commit-Reveal Scheme");
console.log("-".repeat(40));
console.log("Commit to a value without revealing it\n");

// Alice commits to her choice
const aliceChoice = "rock";
const aliceNonce = crypto.getRandomValues(new Uint8Array(32));

const commitment = Keccak256.hashMultiple([
	new TextEncoder().encode(aliceChoice),
	aliceNonce,
]);

console.log("Commit Phase:");
console.log(`Alice's choice: "${aliceChoice}" (kept secret)`);
console.log(`Nonce:          ${Hex.fromBytes(aliceNonce)}`);
console.log(`Commitment:     ${Hex.fromBytes(commitment)}\n`);

// Bob makes his choice (cannot change after seeing commitment)
const bobChoice = "paper";
console.log("Bob sees commitment and makes his choice:");
console.log(`Bob's choice: "${bobChoice}"\n`);

// Reveal Phase
console.log("Reveal Phase:");
const revealed = Keccak256.hashMultiple([
	new TextEncoder().encode(aliceChoice),
	aliceNonce,
]);

const verified = Hex.fromBytes(commitment) === Hex.fromBytes(revealed);
console.log(`Alice reveals: "${aliceChoice}"`);
console.log(`Commitment matches: ${verified}`);
console.log(
	`Winner: ${bobChoice === "paper" && aliceChoice === "rock" ? "Bob" : "Alice"}\n`,
);

// 2. Blind Auction
console.log("2. Blind Auction");
console.log("-".repeat(40));
console.log("Bidders commit to sealed bids, reveal later\n");

interface Bid {
	bidder: string;
	amount: number;
	nonce: Uint8Array;
	commitment: Uint8Array;
}

const bids: Bid[] = [];

// Commit phase
const bidders = [
	{ name: "Alice", amount: 100 },
	{ name: "Bob", amount: 150 },
	{ name: "Carol", amount: 125 },
];

console.log("Commit Phase:");
for (const { name, amount } of bidders) {
	const nonce = crypto.getRandomValues(new Uint8Array(32));
	const amountBytes = new Uint8Array(8);
	new DataView(amountBytes.buffer).setBigUint64(0, BigInt(amount), false);

	const commitment = Keccak256.hashMultiple([
		new TextEncoder().encode(name),
		amountBytes,
		nonce,
	]);

	bids.push({ bidder: name, amount, nonce, commitment });
	console.log(`  ${name}: ${Hex.fromBytes(commitment).slice(0, 20)}...`);
}

console.log("\nReveal Phase:");
let highestBid = 0;
let winner = "";

for (const bid of bids) {
	const amountBytes = new Uint8Array(8);
	new DataView(amountBytes.buffer).setBigUint64(0, BigInt(bid.amount), false);

	const revealed = Keccak256.hashMultiple([
		new TextEncoder().encode(bid.bidder),
		amountBytes,
		bid.nonce,
	]);

	const valid = Hex.fromBytes(bid.commitment) === Hex.fromBytes(revealed);
	console.log(`  ${bid.bidder}: $${bid.amount} (verified: ${valid})`);

	if (valid && bid.amount > highestBid) {
		highestBid = bid.amount;
		winner = bid.bidder;
	}
}

console.log(`\nWinner: ${winner} with bid of $${highestBid}\n`);

// 3. Timestamped Document Commitment
console.log("3. Timestamped Document Commitment");
console.log("-".repeat(40));
console.log("Prove document existed at specific time\n");

const document = "Important legal document content";
const timestamp = Date.now();
const documentHash = Keccak256.hashString(document);

// Create timestamped commitment
const timestampBytes = new Uint8Array(8);
new DataView(timestampBytes.buffer).setBigUint64(0, BigInt(timestamp), false);
const timestampedCommitment = Keccak256.hashMultiple([
	documentHash,
	timestampBytes,
]);

console.log(`Document: "${document.slice(0, 30)}..."`);
console.log(`Timestamp: ${new Date(timestamp).toISOString()}`);
console.log(`Document hash: ${Hex.fromBytes(documentHash)}`);
console.log(`Timestamped commitment: ${Hex.fromBytes(timestampedCommitment)}`);

// Later verification
console.log("\nVerification (later):");
const verifyHash = Keccak256.hashString(document);
const verifyCommitment = Keccak256.hashMultiple([verifyHash, timestampBytes]);

console.log(
	`Document matches: ${Hex.fromBytes(verifyHash) === Hex.fromBytes(documentHash)}`,
);
console.log(
	`Commitment matches: ${Hex.fromBytes(verifyCommitment) === Hex.fromBytes(timestampedCommitment)}\n`,
);

// 4. Merkle Proof
console.log("4. Merkle Proof");
console.log("-".repeat(40));
console.log("Prove inclusion in a set without revealing all items\n");

// Build Merkle tree
const leaves = [
	Keccak256.hashString("Transaction 1"),
	Keccak256.hashString("Transaction 2"),
	Keccak256.hashString("Transaction 3"),
	Keccak256.hashString("Transaction 4"),
];

console.log("Merkle tree leaves:");
for (let i = 0; i < leaves.length; i++) {
	console.log(`  Tx ${i + 1}: ${Hex.fromBytes(leaves[i]).slice(0, 20)}...`);
}

// Build tree
const level1 = [
	Keccak256.hashMultiple([leaves[0], leaves[1]]),
	Keccak256.hashMultiple([leaves[2], leaves[3]]),
];

const root = Keccak256.hashMultiple([level1[0], level1[1]]);
console.log(`\nMerkle root: ${Hex.fromBytes(root)}`);

// Prove Transaction 2 is included
const proofIndex = 1; // Transaction 2
const proof = [
	leaves[0], // Sibling
	level1[1], // Uncle
];

console.log("\nProof for Transaction 2:");
console.log(`  Leaf:    ${Hex.fromBytes(leaves[proofIndex]).slice(0, 20)}...`);
console.log(`  Sibling: ${Hex.fromBytes(proof[0]).slice(0, 20)}...`);
console.log(`  Uncle:   ${Hex.fromBytes(proof[1]).slice(0, 20)}...`);

// Verify proof
const step1 = Keccak256.hashMultiple([proof[0], leaves[proofIndex]]);
const computedRoot = Keccak256.hashMultiple([step1, proof[1]]);

console.log(`\nComputed root: ${Hex.fromBytes(computedRoot)}`);
console.log(
	`Proof valid: ${Hex.fromBytes(computedRoot) === Hex.fromBytes(root)}\n`,
);

// 5. Verifiable Random Function (simplified)
console.log("5. Verifiable Random Function");
console.log("-".repeat(40));
console.log("Generate randomness with proof of correct computation\n");

// Seed-based randomness
const seed = crypto.getRandomValues(new Uint8Array(32));
const input = new TextEncoder().encode("block-123");

console.log(`Seed:  ${Hex.fromBytes(seed)}`);
console.log(`Input: "block-123"`);

// Generate "random" output
const output = Keccak256.hashMultiple([seed, input]);
console.log(`Output: ${Hex.fromBytes(output)}\n`);

// Anyone can verify with same inputs
const verifyOutput = Keccak256.hashMultiple([seed, input]);
console.log("Verification:");
console.log(`  Recomputed: ${Hex.fromBytes(verifyOutput)}`);
console.log(
	`  Matches:    ${Hex.fromBytes(output) === Hex.fromBytes(verifyOutput)}\n`,
);

// 6. Hash Chain / One-Time Passwords
console.log("6. Hash Chain for One-Time Passwords");
console.log("-".repeat(40));

const chainLength = 5;
const initialSecret = crypto.getRandomValues(new Uint8Array(32));

// Build chain backwards
const chain: Uint8Array[] = [initialSecret];
for (let i = 1; i < chainLength; i++) {
	const next = Keccak256.hash(chain[chain.length - 1]);
	chain.push(next);
}

console.log("Hash chain (built backwards):");
for (let i = 0; i < chain.length; i++) {
	console.log(`  Link ${i}: ${Hex.fromBytes(chain[i]).slice(0, 20)}...`);
}

// Publish only the final hash
const publicCommitment = chain[chain.length - 1];
console.log(`\nPublic commitment: ${Hex.fromBytes(publicCommitment)}`);

// Use chain forwards (reveal one at a time)
console.log("\nUsing passwords (forward):");
for (let i = chain.length - 2; i >= 0; i--) {
	const password = chain[i];
	const hashed = Keccak256.hash(password);

	// Verify against previous commitment
	const expectedCommitment = chain[i + 1];
	const valid = Hex.fromBytes(hashed) === Hex.fromBytes(expectedCommitment);

	console.log(
		`  Password ${chain.length - 1 - i}: ${Hex.fromBytes(password).slice(0, 12)}... (valid: ${valid})`,
	);
}

console.log("\nEach password can only be used once");
console.log("Previous passwords cannot be derived from later ones\n");

// 7. Zero-Knowledge Proof (simplified)
console.log("7. Simple Zero-Knowledge Proof");
console.log("-".repeat(40));
console.log("Prove knowledge of secret without revealing it\n");

const secret = "my-secret-password";
const secretHash = Keccak256.hashString(secret);

console.log("Prover commits to secret hash:");
console.log(`  Secret hash: ${Hex.fromBytes(secretHash)}\n`);

// Challenge-response
const challenge = crypto.getRandomValues(new Uint8Array(32));
console.log("Verifier sends challenge:");
console.log(`  Challenge: ${Hex.fromBytes(challenge)}\n`);

// Prover responds with hash(secret || challenge)
const response = Keccak256.hashMultiple([
	new TextEncoder().encode(secret),
	challenge,
]);

console.log("Prover responds:");
console.log(`  Response: ${Hex.fromBytes(response)}\n`);

// Verifier checks (needs to know secret hash, not secret)
// In real ZK proof, verifier wouldn't know secret
const expectedResponse = Keccak256.hashMultiple([
	new TextEncoder().encode(secret),
	challenge,
]);

console.log("Verification:");
console.log(
	`  Valid: ${Hex.fromBytes(response) === Hex.fromBytes(expectedResponse)}`,
);
console.log("  (Prover knows secret without revealing it)\n");

console.log("=== Complete ===");

File: examples/crypto/keccak256/commitment-schemes.zig
const std = @import("std");
const crypto = @import("crypto");

/// Cryptographic Commitment Schemes
///
/// Demonstrates using Keccak256 for commitments:
/// - Commit-reveal schemes
/// - Secret commitments with nonces
/// - Hash chains
/// - Merkle proofs
pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const stdout = std.io.getStdOut().writer();

    try stdout.print("=== Cryptographic Commitment Schemes ===\n\n", .{});

    // 1. Basic Commit-Reveal
    try stdout.print("1. Basic Commit-Reveal Scheme\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});
    try stdout.print("Commit to a value without revealing it\n\n", .{});

    const alice_choice = "rock";
    var alice_nonce: [32]u8 = undefined;
    std.crypto.random.bytes(&alice_nonce);

    // Create commitment: hash(choice || nonce)
    var commit_data = std.ArrayList(u8).init(allocator);
    defer commit_data.deinit();
    try commit_data.appendSlice(alice_choice);
    try commit_data.appendSlice(&alice_nonce);

    var commitment: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(commit_data.items, &commitment);

    try stdout.print("Commit Phase:\n", .{});
    try stdout.print("Alice's choice: \"{s}\" (kept secret)\n", .{alice_choice});
    try stdout.print("Nonce:          0x", .{});
    for (alice_nonce) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("\nCommitment:     0x", .{});
    for (commitment) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("\n\n", .{});

    const bob_choice = "paper";
    try stdout.print("Bob sees commitment and makes his choice:\n", .{});
    try stdout.print("Bob's choice: \"{s}\"\n\n", .{bob_choice});

    // Reveal Phase
    try stdout.print("Reveal Phase:\n", .{});
    var revealed: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(commit_data.items, &revealed);

    const verified = std.mem.eql(u8, &commitment, &revealed);
    try stdout.print("Alice reveals: \"{s}\"\n", .{alice_choice});
    try stdout.print("Commitment matches: {}\n", .{verified});

    const bob_wins = std.mem.eql(u8, bob_choice, "paper") and std.mem.eql(u8, alice_choice, "rock");
    try stdout.print("Winner: {s}\n\n", .{if (bob_wins) "Bob" else "Alice"});

    // 2. Timestamped Document Commitment
    try stdout.print("2. Timestamped Document Commitment\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});
    try stdout.print("Prove document existed at specific time\n\n", .{});

    const document = "Important legal document content";
    const timestamp = std.time.milliTimestamp();

    var doc_hash: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(document, &doc_hash);

    // Create timestamped commitment
    var ts_data = std.ArrayList(u8).init(allocator);
    defer ts_data.deinit();
    try ts_data.appendSlice(&doc_hash);

    var ts_bytes: [8]u8 = undefined;
    std.mem.writeInt(u64, &ts_bytes, @bitCast(timestamp), .big);
    try ts_data.appendSlice(&ts_bytes);

    var ts_commitment: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(ts_data.items, &ts_commitment);

    try stdout.print("Document: \"{s}...\"\n", .{document[0..30]});
    try stdout.print("Timestamp: {d}\n", .{timestamp});
    try stdout.print("Document hash: 0x", .{});
    for (doc_hash) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("\nTimestamped commitment: 0x", .{});
    for (ts_commitment) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("\n\n", .{});

    // 3. Merkle Proof
    try stdout.print("3. Merkle Proof\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});
    try stdout.print("Prove inclusion in a set without revealing all items\n\n", .{});

    // Build Merkle tree
    var leaves: [4][32]u8 = undefined;
    try crypto.keccak_asm.keccak256("Transaction 1", &leaves[0]);
    try crypto.keccak_asm.keccak256("Transaction 2", &leaves[1]);
    try crypto.keccak_asm.keccak256("Transaction 3", &leaves[2]);
    try crypto.keccak_asm.keccak256("Transaction 4", &leaves[3]);

    try stdout.print("Merkle tree leaves:\n", .{});
    for (leaves, 0..) |leaf, i| {
        try stdout.print("  Tx {d}: 0x", .{i + 1});
        for (leaf[0..10]) |byte| try stdout.print("{x:0>2}", .{byte});
        try stdout.print("...\n", .{});
    }

    // Build tree level 1
    var level1: [2][32]u8 = undefined;
    var node_data: [64]u8 = undefined;

    @memcpy(node_data[0..32], &leaves[0]);
    @memcpy(node_data[32..64], &leaves[1]);
    try crypto.keccak_asm.keccak256(&node_data, &level1[0]);

    @memcpy(node_data[0..32], &leaves[2]);
    @memcpy(node_data[32..64], &leaves[3]);
    try crypto.keccak_asm.keccak256(&node_data, &level1[1]);

    // Build root
    @memcpy(node_data[0..32], &level1[0]);
    @memcpy(node_data[32..64], &level1[1]);
    var root: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(&node_data, &root);

    try stdout.print("\nMerkle root: 0x", .{});
    for (root) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("\n\n", .{});

    // Prove Transaction 2 is included
    const proof_index: usize = 1;
    try stdout.print("Proof for Transaction 2:\n", .{});
    try stdout.print("  Leaf:    0x", .{});
    for (leaves[proof_index][0..10]) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("...\n  Sibling: 0x", .{});
    for (leaves[0][0..10]) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("...\n  Uncle:   0x", .{});
    for (level1[1][0..10]) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("...\n\n", .{});

    // Verify proof
    @memcpy(node_data[0..32], &leaves[0]);
    @memcpy(node_data[32..64], &leaves[proof_index]);
    var step1: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(&node_data, &step1);

    @memcpy(node_data[0..32], &step1);
    @memcpy(node_data[32..64], &level1[1]);
    var computed_root: [32]u8 = undefined;
    try crypto.keccak_asm.keccak256(&node_data, &computed_root);

    try stdout.print("Computed root: 0x", .{});
    for (computed_root) |byte| try stdout.print("{x:0>2}", .{byte});
    const proof_valid = std.mem.eql(u8, &computed_root, &root);
    try stdout.print("\nProof valid: {}\n\n", .{proof_valid});

    // 4. Hash Chain for One-Time Passwords
    try stdout.print("4. Hash Chain for One-Time Passwords\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    const chain_length = 5;
    var chain = try allocator.alloc([32]u8, chain_length);
    defer allocator.free(chain);

    // Initial secret
    std.crypto.random.bytes(&chain[0]);

    // Build chain forwards
    for (1..chain_length) |i| {
        try crypto.keccak_asm.keccak256(&chain[i - 1], &chain[i]);
    }

    try stdout.print("Hash chain (built forwards):\n", .{});
    for (chain, 0..) |link, i| {
        try stdout.print("  Link {d}: 0x", .{i});
        for (link[0..10]) |byte| try stdout.print("{x:0>2}", .{byte});
        try stdout.print("...\n", .{});
    }

    // Publish only the final hash
    try stdout.print("\nPublic commitment: 0x", .{});
    for (chain[chain_length - 1]) |byte| try stdout.print("{x:0>2}", .{byte});
    try stdout.print("\n\n", .{});

    // Use chain backwards (reveal one at a time)
    try stdout.print("Using passwords (backwards):\n", .{});
    var i: usize = chain_length - 1;
    while (i > 0) : (i -= 1) {
        const password = &chain[i - 1];
        var hashed: [32]u8 = undefined;
        try crypto.keccak_asm.keccak256(password, &hashed);

        const expected_commitment = &chain[i];
        const valid = std.mem.eql(u8, &hashed, expected_commitment);

        try stdout.print("  Password {d}: 0x", .{chain_length - i});
        for (password[0..6]) |byte| try stdout.print("{x:0>2}", .{byte});
        try stdout.print("... (valid: {})\n", .{valid});
    }

    try stdout.print("\nEach password can only be used once\n", .{});
    try stdout.print("Previous passwords cannot be derived from later ones\n\n", .{});

    // 5. Blind Commitment
    try stdout.print("5. Blind Commitment (Sealed Bid)\n", .{});
    try stdout.print("{s}\n", .{"-" ** 40});

    const Bid = struct {
        bidder: []const u8,
        amount: u64,
        nonce: [32]u8,
        commitment: [32]u8,
    };

    var bids = std.ArrayList(Bid).init(allocator);
    defer bids.deinit();

    const bidders = [_]struct { name: []const u8, amount: u64 }{
        .{ .name = "Alice", .amount = 100 },
        .{ .name = "Bob", .amount = 150 },
        .{ .name = "Carol", .amount = 125 },
    };

    try stdout.print("Commit Phase:\n", .{});
    for (bidders) |bidder| {
        var bid: Bid = undefined;
        bid.bidder = bidder.name;
        bid.amount = bidder.amount;
        std.crypto.random.bytes(&bid.nonce);

        var bid_data = std.ArrayList(u8).init(allocator);
        defer bid_data.deinit();
        try bid_data.appendSlice(bidder.name);

        var amount_bytes: [8]u8 = undefined;
        std.mem.writeInt(u64, &amount_bytes, bidder.amount, .big);
        try bid_data.appendSlice(&amount_bytes);
        try bid_data.appendSlice(&bid.nonce);

        try crypto.keccak_asm.keccak256(bid_data.items, &bid.commitment);

        try stdout.print("  {s}: 0x", .{bidder.name});
        for (bid.commitment[0..10]) |byte| try stdout.print("{x:0>2}", .{byte});
        try stdout.print("...\n", .{});

        try bids.append(bid);
    }

    try stdout.print("\nReveal Phase:\n", .{});
    var highest_bid: u64 = 0;
    var winner: []const u8 = "";

    for (bids.items) |bid| {
        var bid_data = std.ArrayList(u8).init(allocator);
        defer bid_data.deinit();
        try bid_data.appendSlice(bid.bidder);

        var amount_bytes: [8]u8 = undefined;
        std.mem.writeInt(u64, &amount_bytes, bid.amount, .big);
        try bid_data.appendSlice(&amount_bytes);
        try bid_data.appendSlice(&bid.nonce);

        var revealed: [32]u8 = undefined;
        try crypto.keccak_asm.keccak256(bid_data.items, &revealed);

        const valid = std.mem.eql(u8, &bid.commitment, &revealed);
        try stdout.print("  {s}: ${d} (verified: {})\n", .{ bid.bidder, bid.amount, valid });

        if (valid and bid.amount > highest_bid) {
            highest_bid = bid.amount;
            winner = bid.bidder;
        }
    }

    try stdout.print("\nWinner: {s} with bid of ${d}\n\n", .{ winner, highest_bid });

    try stdout.print("=== Complete ===\n", .{});
}
